File: 卡码/代码随想录/Python/贪心/02-中等/LeetCode_0738.py
#738. 单调递增的数字
from typing import *

class Solution:
  def monotoneIncreasingDigits(self, n: int) -> int:
    nums:List[int]=list(str(n))
    start:int=len(nums)
    for i in range(len(nums)-1,0,-1):
      if nums[i-1]>nums[i]:
        nums[i-1]=str(int(nums[i-1])-1)
        start=i
    for i in range(start,len(nums)):
      nums[i]=9
    nums="".join(nums[:start])+'9'*(len(nums)-start)
    return int(nums)

File: 卡码/代码随想录/1/08-贪心算法/Java/greedy/src/main/java/com/kurumi/greedy/_02_中等/LeetCode_0738.java
//738. 单调递增的数字
package com.kurumi.greedy._02_中等;

public class LeetCode_0738 {

}
class Solution {
  public int monotoneIncreasingDigits(int n) {
    char[] nums = String.valueOf(n).toCharArray();
    int start=nums.length;
    for(int i=nums.length-1;i>0;i--){
      if(nums[i-1]>nums[i]){
        nums[i-1]--;
        start=i;
      }
    }
    for(int i=start;i<nums.length;i++) nums[i]='9';
    return Integer.parseInt(String.valueOf(nums));
  }
}
